### PR TITLE
empty graph bugfix

### PR DESCRIPTION
**Pull request by davidegraff**
## Description
This PR fixes a bug in the new `torch_scatter` backend introduced in #470 in the case of empty graphs having no prediction and being silently squashed.

## Previous treatment
In `v1` we treated empty graphs with a zero vector, so we should do the same here.

## Bugfix
During `MolGraph` preparation we now check if a graph is empty `n_atoms == 0` and produce a vertex feature of zeros of shape `1 x atom_fdim`. This _should_ produce a latent vector of zeros if not for the fact that the output matrix $\mathbf W_o$ used in message passing has `bias=True` and this is _not_ controlled by the input `bias` argument. This was the case in v1 so it's the same here, but we could think about changing this in the future. As such, the latent vector for an empty graph will always be equal to the learned bias term $\mathbf b$ of the output matrix.

## NOTE
- we should add a unit test for this but I don't have the time.

## PR Stack
- #514
- #470
- #552 
- #536 
- #537 <-- HERE

### CODE REVIEW

File: chemprop/v2/featurizers/molgraph/molecule.py [hunk 1]
```diff
@@ -90,42 +87,35 @@ def __call__(
                 f"got: {n_bonds} and {len(bond_features_extra)}, respectively"
             )
 
-        X_v = np.array([self.atom_featurizer(a) for a in mol.GetAtoms()])
-        X_e = np.empty((2 * n_bonds, self.bond_fdim))
-        a2b = [[] for _ in range(n_atoms)]
-        b2a = np.empty(2 * n_bonds, int)
-        b2revb = np.empty(2 * n_bonds, int)
+        if n_atoms == 0:
+            V = np.zeros((0, self.atom_fdim))
```
Comment (davidegraff) [inline chemprop/v2/featurizers/molgraph/molecule.py:91]:
```suggestion
            V = np.zeros((1, self.atom_fdim))
```

**Review by kevingreenman — APPROVED** [review]

**Comment by KnathanM** [comment]
As I am looking at the bond featurizing in #931, I wondered why we need to check `if n_atoms == 0`, which led me here. When would Chemprop be given an empty graph? To me an empty graph shouldn't have a target value to begin with. 

I also couldn't find where in the v1 code empty graphs get featurized to zero vectors. A pointer to that would be appreciated. 

I do get an empty graph when I do `mol = rdkit.Chem.MolFromSmiles("")` and could use that to write a unit test for this. But I don't think using an empty string makes sense for a SMILES.

**Comment by davidegraff** [comment]
It's implicit logic in the v1 code, that if you passed in a batch of 50 graphs with one empty graph, you would still generate 50 predictions with one of them originating from the $\mathbf 0$ vector in the latent space. You'd have to look back through the batching and scattering logic to reason this out.

**Comment by KnathanM** [comment]
Okay, without checking it myself, it seems reasonable that the batching and scattering logic of v1 would do this. 
My follow up questions is why do we want Chemprop to do this? I think an empty graph is more indicative of an error on the user's part than an actual datapoint. The only empty graph I could find is `rdkit.Chem.MolFromSmiles("")`, which I think is nonsense input. Originally I thought `rdkit.Chem.MolFromSMiles("[H]").GetNumAtoms()` would return 0, but rdkit doesn't remove hydrogen atoms without neighbors. 
Removing the check `if n_atoms == 0` would be a small simplification to the code.
To me the larger issue is if Chemprop should treat empty graphs like errors and crash, or silently accept the nonsense input and move on.

**Comment by davidegraff** [comment]
It's an inductive bias placed on the model. You can argue it doesn't make sense, but it does impart _some_ structure to the latent space so that all empty graphs are encoded to $\mathbf 0$. Now whether you should actually see these examples during training and whether they're even valid examples is another thing entirely. It's a larger question to answer and would represent a small but significant algorithmic change.

**Pull request merged**